File: reverseInParentheses.go
package main

import "fmt"

func main() {
	fmt.Println(solution("foo(bar(baz))blim"))
}

func solution(inputString string) string {
	output, _ := getStringPortion(inputString, 0, false)
	return string(output)
}

func getStringPortion(inputString string, startIndex int, reverse bool) ([]rune, int) {
	portion := []rune{}
	for index := startIndex; index < len(inputString); index++ {
		char := inputString[index]
		if char == ')' {
			return portion, index
		}
		newPortion := []rune{}
		if char == '(' {
			newPortion, index = getStringPortion(inputString, index+1, !reverse)
		} else {
			newPortion = []rune{rune(char)}
		}
		if reverse {
			portion = append(newPortion, portion...)
		} else {
			portion = append(portion, newPortion...)
		}
	}
	return portion, len(inputString)
}

File: reverseInParentheses.py
def solution(input_string):
    output, _ = get_string_portion(input_string, 0, False)
    return output


def get_string_portion(input_string, start_index, reverse):
    portion = ""
    index = start_index
    while index < len(input_string):
        char = input_string[index]
        if char == ')':
            return portion, index
        if char == '(':
            new_portion, index = get_string_portion(input_string, index + 1, not reverse)
        else:
            new_portion = char
        if reverse:
            portion = new_portion + portion
        else:
            portion = portion + new_portion
        index += 1
    return portion, len(input_string)


print(solution("foo(bar(baz))blim"))
